### ISSUE TITLE
Refactor show command logic in parser to encode the optionality of in_cluster to the type of object 

### ISSUE THREAD

**Issue by chaas** [MaterializeInc/materialize#15479]
Only materialized views and indexes are associated with clusters. Therefore, we can only support the IN CLUSTER filter to SHOW commands on those object types.
Currently, the parser code uses a if statement to check the object type is a materialized view before parsing the `IN CLUSTER` keyword:
https://github.com/chaas/materialize/blob/31de340184420d84a1073d6457d5d5abe78b1216/src/sql-parser/src/parser.rs#L4606-L4611

Refactor the parser to encode the optionality of in_cluster to the type of object under inspection in the type system. In an ideal world, we'd have something like:

```
enum ShowObjectType {
    MaterializedView { in_cluster: Option<String> },
    Index { in_cluster: Option<String> },
    Table,
}
```
(Suggestion taken from https://github.com/MaterializeInc/materialize/pull/15401#pullrequestreview-1141512948 by @benesch )

**Comment by CNLHC** [comment]
I just submit a PR for this issue https://github.com/MaterializeInc/materialize/pull/16437

I am learning this project, please forgive me if I mistakenly understand the problem.

**Comment by benesch** [comment]
Fixed in #16437. Thanks, @CNLHC!
